### PR TITLE
test(mme): Add injection of state loaded in SPGW state manager (issue #9606) 

### PR DESCRIPTION
**Pull request by thhung**
Signed-off-by: Hung NGUYEN <nguyen.thanh.hung.91@gmail.com>

## Summary

Adding injection of state loaded in SPGW state manager by adding a method to "load" predefined state on the MME task - SPGW, mocking the read_ue_state_from_db() function.
## Test Plan
This change should pass the tests in test_spgw_procedure

## Additional Information

- [ ] This change is backwards-breaking

### CODE REVIEW

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 1]
```diff
@@ -0,0 +1,49 @@
+#include <fstream>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp:13]: For all new files added, please add Magma copyright header (you can pick it up from already added files)

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h [hunk 2]
```diff
@@ -0,0 +1,19 @@
+#include <iostream>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h:13]: add copyright header

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 3]
```diff
@@ -0,0 +1,196 @@
+#include <gtest/gtest.h>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp:18]: add copyright header

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    std::cerr << "couldn't open file: " << file_name << std::endl;
+  }
+
+  file_content.close();
+  return vector_of_lines;
+}
+
+status_code_e mock_read_spgw_ue_state_db(
+    const std::vector<std::string>& ue_samples) {
+  for (auto name_of_sample : ue_samples) {
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp]:
```suggestion
  for (const auto& name_of_sample : ue_samples) {
```

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+    spgw_ue_context_t* ue_context_p =
+        (spgw_ue_context_t*) calloc(1, sizeof(spgw_ue_context_t));
+    SpgwStateConverter::proto_to_ue(ue_proto, ue_context_p);
+  }
+  return RETURNok;
+}
+
+}  // namespace lte
+}  // namespace magma
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp]: add blank newline

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+#define DEFAULT_SPGW_CONTEXT_DATA_PATH                                         \
+  "/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/data/"
+
+std::vector<std::string> load_file_into_vector_of_line_content(
+    const std::string& data_folder_path, const std::string& file_name);
+status_code_e mock_read_spgw_ue_state_db(
+    const std::vector<std::string>& ue_samples);
+}  // namespace lte
+}  // namespace magma
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h]: add blank newline at the end of file

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h [hunk 7]
```diff
@@ -0,0 +1,19 @@
...
+#include <string>
+#include <vector>
+
+#include "spgw_state_converter.h"
+
+namespace magma {
+namespace lte {
+
+#define DEFAULT_SPGW_CONTEXT_DATA_PATH                                         \
+  "/home/vagrant/magma/lte/gateway/c/core/oai/test/spgw_task/data/"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h]: I would suggest getting the user's home directory instead and prepend it instead of harcoding `/home/vagrant/`
```
    homedir = getpwuid(getuid())->pw_dir;
```

(or better yet, can we get the relative path??)

**Review by ardzoht — CHANGES_REQUESTED** [review]

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#include <fstream>
+#include <algorithm>
+
+#include "mock_spgw_op.h"
+#include "spgw_state.h"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp]:
```suggestion
#include "lte/gateway/c/core/oai/include/spgw_state.h"
```

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+#include <string>
+#include <thread>
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <pwd.h>
+
+#include "../mock_tasks/mock_tasks.h"
+#include "spgw_test_util.h"
+#include "spgw_state.h"
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]:
```suggestion
#include "lte/gateway/c/core/oai/include/spgw_state.h"
```

Comment (ardzoht) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Please also mark the rest of includes with "lte/gateway/..." full path

**Review by ardzoht — COMMENTED** [review]
@thhung left couple of more comments related with include headers path, after you resolve this the PR can be marked ready for review, thanks for the updates!

**Comment by ardzoht** [comment]
@thhung You need to rebase your branch on top of master to resolve merge conflicts

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 10]
```diff
@@ -0,0 +1,216 @@
...
+#include <sys/types.h>
+#include <pwd.h>
+
+#include "lte/gateway/c/core/oai/test/mock_tasks/mock_tasks.h"
+#include "lte/gateway/c/core/oai/test/spgw_task/spgw_test_util.h"
+#include "lte/gateway/c/core/oai/include/spgw_state.h"
+#include "lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.h"
+#include "lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h"
+
+extern "C" {
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp:27]: these should be before cpp-exclusive headers

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 11]
```diff
@@ -0,0 +1,216 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <gtest/gtest.h>
+#include <string>
+#include <thread>
+
+#include <unistd.h>
+#include <sys/types.h>
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp:15]: can you take care of the cpplint comments on this file? looks like the order is mixed up

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 12]
```diff
@@ -0,0 +1,216 @@
...
+        TASK_MAIN, task_id_list, 2, handle_message, &task_zmq_ctx_main_spgw);
+
+    std::thread task_mme_app(start_mock_mme_app_task, mme_app_handler);
+    task_mme_app.detach();
+
+    std::cout << "Running setup" << std::endl;
+    // initialize the SPGW task
+    spgw_app_init(&spgw_config, mme_config.use_stateless);
+
+    std::string homedir = std::string(getpwuid(getuid())->pw_dir);
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: I'd recommend adding a comment here to specify that the state injection is applied here

**Review by ardzoht — APPROVED** [review]
lgtm, left some minor comments, please take care of them

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+
+#include "lte/gateway/c/core/oai/tasks/sgw/spgw_state_converter.h"
+
+namespace magma {
+namespace lte {
+
+#define DEFAULT_SPGW_CONTEXT_DATA_PATH                                         \
+  "magma/lte/gateway/c/core/oai/test/spgw_task/data/"
+
+std::vector<std::string> load_file_into_vector_of_line_content(
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.h:26]: Please add a brief description of each of these functions

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 14]
```diff
@@ -0,0 +1,218 @@
...
+      &spgw_eps_bearer_ctxt_info_p->sgw_eps_bearer_context_information
+           .pdn_connection,
+      DEFAULT_EPS_BEARER_ID);
+
+  // send an IP alloc response to SPGW
+  itti_ip_allocation_response_t test_ip_alloc_resp = {};
+  fill_ip_allocation_response(
+      &test_ip_alloc_resp, SGI_STATUS_OK, ue_sgw_teid, DEFAULT_EPS_BEARER_ID,
+      test_ue_ip2, DEFAULT_VLAN);
+  status_code_e ip_alloc_rc = sgw_handle_ip_allocation_rsp(
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Is the state for `test_imsi64`  captured before IP allocation has happened, i.e. before Attach was complete? 
If yes, then the same state would not work for TestDeleteSessionSuccess.
If not, then SPGW task should not expect an IP allocation response for this IMSI.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 15]
```diff
@@ -0,0 +1,218 @@
...
+      &sample_delete_session_request, test_mme_s11_teid, ue_sgw_teid,
+      DEFAULT_EPS_BEARER_ID, test_plmn);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_delete_sess_rsp()).Times(1);
+
+  status_code_e return_code = RETURNerror;
+  return_code               = sgw_handle_delete_session_request(
+      &sample_delete_session_request, test_imsi64);
+  ASSERT_EQ(return_code, RETURNok);
+
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp:621]: Please add a check here that number of sessions has reduced by 1 after delete session succeeded, i.e.

`ASSERT_TRUE(is_num_sessions_valid(test_imsi64, name_of_ue_samples.size()-1, 0));`

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 16]
```diff
@@ -0,0 +1,218 @@
...
+  // send an IP alloc response to SPGW
+  itti_ip_allocation_response_t test_ip_alloc_resp = {};
+  fill_ip_allocation_response(
+      &test_ip_alloc_resp, SGI_STATUS_OK, ue_sgw_teid, DEFAULT_EPS_BEARER_ID,
+      test_ue_ip2, DEFAULT_VLAN);
+  status_code_e ip_alloc_rc = sgw_handle_ip_allocation_rsp(
+      spgw_state, &test_ip_alloc_resp, test_imsi64);
+
+  // check if IP address is allocated after this message is done
+  ASSERT_TRUE(eps_bearer_ctxt_p->paa.ipv4_address.s_addr == test_ue_ip2);
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Can you check that this assert is false before the call to `sgw_handle_ip_allocation_rsp`?

**Review by ssanadhya — CHANGES_REQUESTED** [review]
@thhung , could you please add the UE state in the data file names too? For example, are all the UEs in registered state or right after security mode complete?

Further, this PR is only testing for one IMSI. There is no need to add user data for 100 IMSIs with this PR. Even if you want to check multi UE state read, please add 2-3 UEs. More UE data can be added with future PRs.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp (hunk 14 above)
Comment (thhung) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Hi @ssanadhya , the state for the `test_imsi64` captured after Attach was completed. I will update the test.

Comment (thhung) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: I removed this step in my test. Thanks.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp (hunk 16 above)
Comment (thhung) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: @ssanadhya Correct me if I am wrong. Since the state for the test_imsi64 captured after Attach was completed, SPGW task should not expect an IP allocation response for this IMSI, then we don't need to check that this assert is false anymore. Is this correct?

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp (hunk 15 above)
Comment (thhung) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp:621]: @ssanadhya  Would it be fine if I add this check after I create a PR to split `is_num_sessions_valid()` into 2 functions? 

- `is_num_ue_contexts_valid()`
- `is_num_teids_valid()` 

This dues to the fact that after deleting the current UE context, `ue_context_p` is not valid anymore. So this check cannot be passed.  
Otherwise, we could check if `ue_context_p` is valid or not. This should be enough for this check.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp (hunk 16 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: @thhung , that's right. In fact, since the state was captured after Attach was complete, the TestPCEFFailure is not applicable here, since session is already created. You can add tests for 
- Delete Session
- Release Bearer
- Secondary PDN connections 
- Dedicated Bearer activation/deactivation

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp (hunk 15 above)
Comment (ssanadhya) [reply lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp:621]: Yes, add the check for valid ue_context_p here, then split `is_num_sessions_valid()` in a separate PR.

File: lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.c [hunk 17]
```diff
@@ -1006,13 +1006,15 @@ status_code_e sgw_handle_modify_bearer_request(
...
               eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up,
               eps_bearer_ctxt_p->eps_bearer_id);
-          // This is best effort, ignore return code.
-          gtp_tunnel_ops->send_end_marker(enb, modify_bearer_pP->teid);
-          // delete GTPv1-U tunnel
-          rv = gtp_tunnel_ops->del_tunnel(
-              enb, enb_ipv6, ue_ipv4, ue_ipv6,
-              eps_bearer_ctxt_p->s_gw_teid_S1u_S12_S4_up,
-              eps_bearer_ctxt_p->enb_teid_S1u, NULL);
+          if (gtp_tunnel_ops) {
```
Comment (ardzoht) [inline lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.c:1009]: what is this change needed for? is `gtp_tunnel_ops` not valid on the unit tests?

Comment (thhung) [reply lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.c:1009]: Yes, in one of these tests, gtp_tunnel_ops is not initialized. Therefore, I added this check to make sure the safety. I am asking @ssanadhya for her feedback. Let's see if we have a better way.

Comment (ssanadhya) [reply lte/gateway/c/core/oai/tasks/sgw/sgw_handlers.c:1009]: @thhung , I have used the `#if !MME_UNIT_TEST` directive to skip calls to gtp_tunnel_ops for unit testing. Please use the same here.

I have a work in progress PR https://github.com/magma/magma/pull/10702 to create a mock GTP tunnel for unit testing. In the mean time let's use `#if !MME_UNIT_TEST`.

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 18]
```diff
@@ -0,0 +1,63 @@
...
+
+namespace magma {
+namespace lte {
+
+// loads paths of data samples of spgw states (storing in each line) from a file
+std::vector<std::string> load_file_into_vector_of_line_content(
+    const std::string& data_folder_path, const std::string& file_name) {
+  std::fstream file_content(file_name.c_str(), std::ios_base::in);
+  std::string line;
+  std::vector<std::string> vector_of_lines;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp]:
```suggestion
  std::vector<std::string> vector_of_file_names;
```

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 19]
```diff
@@ -0,0 +1,63 @@
...
+#include <google/protobuf/text_format.h>
+
+namespace magma {
+namespace lte {
+
+// loads paths of data samples of spgw states (storing in each line) from a file
+std::vector<std::string> load_file_into_vector_of_line_content(
+    const std::string& data_folder_path, const std::string& file_name) {
+  std::fstream file_content(file_name.c_str(), std::ios_base::in);
+  std::string line;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp]:
```suggestion
  std::string data_file_name;
```

File: lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp [hunk 20]
```diff
@@ -0,0 +1,63 @@
...
+
+  file_content.close();
+  return vector_of_lines;
+}
+
+// mocking the reading spgw ue states from redis database by injecting local
+// samples
+status_code_e mock_read_spgw_ue_state_db(
+    const std::vector<std::string>& ue_samples) {
+  for (const auto& name_of_sample : ue_samples) {
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/mock_spgw_op.cpp]:
```suggestion
  for (const auto& name_of_sample_file : ue_samples) {
```

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 21]
```diff
@@ -0,0 +1,719 @@
...
+  fill_modify_bearer_request(
+      &sample_modify_bearer_req, DEFAULT_MME_S11_TEID, ue_sgw_teid,
+      DEFAULT_ENB_GTP_TEID, DEFAULT_BEARER_INDEX, DEFAULT_EPS_BEARER_ID);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_modify_bearer_rsp()).Times(1);
+  status_code_e return_code =
+      sgw_handle_modify_bearer_request(&sample_modify_bearer_req, test_imsi64);
+
+  ASSERT_EQ(return_code, RETURNok);
+
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Since the UE is attached, there will not be any modify bearer request. Please remove this.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 22]
```diff
@@ -0,0 +1,719 @@
...
+  fill_modify_bearer_request(
+      &sample_modify_bearer_req, test_mme_s11_teid, ue_sgw_teid,
+      DEFAULT_ENB_GTP_TEID, test_bearer_index, DEFAULT_EPS_BEARER_ID);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_modify_bearer_rsp()).Times(1);
+  return_code =
+      sgw_handle_modify_bearer_request(&sample_modify_bearer_req, test_imsi64);
+
+  ASSERT_EQ(return_code, RETURNok);
+
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Not needed here

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 23]
```diff
@@ -0,0 +1,719 @@
...
+TEST_F(SPGWAppInjectedStateProcedureTest, TestDedicatedBearerDeactivation) {
+  spgw_state_t* spgw_state  = get_spgw_state(false);
+  status_code_e return_code = RETURNerror;
+
+  spgw_ue_context_t* ue_context_p = spgw_get_ue_context(test_imsi64);
+  teid_t ue_sgw_teid =
+      LIST_FIRST(&ue_context_p->sgw_s11_teid_list)->sgw_s11_teid;
+
+  // Verify that no IP address is allocated for this UE
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Irrelevant comment

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 24]
```diff
@@ -0,0 +1,719 @@
...
+  fill_modify_bearer_request(
+      &sample_modify_bearer_req, test_mme_s11_teid, ue_sgw_teid,
+      DEFAULT_ENB_GTP_TEID, DEFAULT_BEARER_INDEX, DEFAULT_EPS_BEARER_ID);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_modify_bearer_rsp()).Times(1);
+  return_code =
+      sgw_handle_modify_bearer_request(&sample_modify_bearer_req, test_imsi64);
+
+  EXPECT_EQ(return_code, RETURNok);
+
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Remove

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 25]
```diff
@@ -0,0 +1,719 @@
...
+  fill_modify_bearer_request(
+      &sample_modify_bearer_req, test_mme_s11_teid, ue_sgw_teid,
+      DEFAULT_ENB_GTP_TEID, DEFAULT_BEARER_INDEX, DEFAULT_EPS_BEARER_ID);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_modify_bearer_rsp()).Times(1);
+  return_code =
+      sgw_handle_modify_bearer_request(&sample_modify_bearer_req, test_imsi64);
+
+  ASSERT_EQ(return_code, RETURNok);
+
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Remove

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 26]
```diff
@@ -0,0 +1,719 @@
...
+  fill_modify_bearer_request(
+      &sample_modify_bearer_req, test_mme_s11_teid, ue_sgw_teid,
+      DEFAULT_ENB_GTP_TEID, DEFAULT_BEARER_INDEX, DEFAULT_EPS_BEARER_ID);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_modify_bearer_rsp()).Times(1);
+  return_code =
+      sgw_handle_modify_bearer_request(&sample_modify_bearer_req, test_imsi64);
+
+  EXPECT_EQ(return_code, RETURNok);
+
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Remove

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 27]
```diff
@@ -0,0 +1,719 @@
...
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Remove

**Review by ssanadhya — CHANGES_REQUESTED** [review]
@thhung , if you are only using 3 data files in data_list.txt, then you don't need to commit the data file for all 100+ IMSIs in this PR. Could you please trim this PR to only commit the relevant UE data files?

**Comment by thhung** [comment]
@ssanadhya I removed all irrelevant UE data files.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 28]
```diff
@@ -0,0 +1,646 @@
...
+    // Sleep to ensure that messages are received and contexts are released
+    std::this_thread::sleep_for(
+        std::chrono::milliseconds(END_OF_TEST_SLEEP_MS));
+  }
+
+ protected:
+  std::shared_ptr<MockMmeAppHandler> mme_app_handler;
+  std::string test_imsi_str = "001010000000002";
+  uint64_t test_imsi64      = 1010000000002;
+  uint64_t test_imsi64_test = 1010000000001;
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]:
```suggestion
  uint64_t test_imsi64_invalid = 1010000000001;
```

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 29]
```diff
@@ -0,0 +1,646 @@
...
+TEST_F(SPGWAppInjectedStateProcedureTest, TestReleaseBearerWithInvalidImsi64) {
+  spgw_state_t* spgw_state        = get_spgw_state(false);
+  status_code_e return_code       = RETURNerror;
+  spgw_ue_context_t* ue_context_p = spgw_get_ue_context(test_imsi64);
+  teid_t ue_sgw_teid =
+      LIST_FIRST(&ue_context_p->sgw_s11_teid_list)->sgw_s11_teid;
+  s_plus_p_gw_eps_bearer_context_information_t* spgw_eps_bearer_ctxt_info_p =
+      sgw_cm_get_spgw_context(ue_sgw_teid);
+
+  sgw_eps_bearer_ctxt_t* eps_bearer_ctxt_p = sgw_cm_get_eps_bearer_entry(
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: `eps_bearer_ctxt_p`is not getting used, please remove

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 30]
```diff
@@ -0,0 +1,646 @@
...
+  status_code_e return_code = RETURNerror;
+
+  spgw_ue_context_t* ue_context_p = spgw_get_ue_context(test_imsi64);
+  teid_t ue_sgw_teid =
+      LIST_FIRST(&ue_context_p->sgw_s11_teid_list)->sgw_s11_teid;
+
+  s_plus_p_gw_eps_bearer_context_information_t* spgw_eps_bearer_ctxt_info_p =
+      sgw_cm_get_spgw_context(ue_sgw_teid);
+
+  sgw_eps_bearer_ctxt_t* eps_bearer_ctxt_p = sgw_cm_get_eps_bearer_entry(
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Unused, please remove

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 31]
```diff
@@ -0,0 +1,646 @@
...
+  spgw_state_t* spgw_state  = get_spgw_state(false);
+  status_code_e return_code = RETURNerror;
+
+  spgw_ue_context_t* ue_context_p = spgw_get_ue_context(test_imsi64);
+  teid_t ue_sgw_teid =
+      LIST_FIRST(&ue_context_p->sgw_s11_teid_list)->sgw_s11_teid;
+  s_plus_p_gw_eps_bearer_context_information_t* spgw_eps_bearer_ctxt_info_p =
+      sgw_cm_get_spgw_context(ue_sgw_teid);
+
+  sgw_eps_bearer_ctxt_t* eps_bearer_ctxt_p = sgw_cm_get_eps_bearer_entry(
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Unused, please remove

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 32]
```diff
@@ -0,0 +1,646 @@
...
+      DEFAULT_ENB_GTP_TEID, DEFAULT_BEARER_INDEX, DEFAULT_EPS_BEARER_ID);
+
+  EXPECT_CALL(*mme_app_handler, mme_app_handle_modify_bearer_rsp()).Times(1);
+  return_code =
+      sgw_handle_modify_bearer_request(&sample_modify_bearer_req, test_imsi64);
+
+  ASSERT_EQ(return_code, RETURNok);
+
+  // verify that sessions exist in SPGW state
+  ASSERT_TRUE(is_num_sessions_valid(test_imsi64, name_of_ue_samples.size(), 1));
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Add this check before sending the modify bearer request too. That is, the number of valid sessions should not change before and after the modify bearer request.

File: lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp [hunk 33]
```diff
@@ -0,0 +1,646 @@
...
+  std::this_thread::sleep_for(std::chrono::milliseconds(END_OF_TEST_SLEEP_MS));
+}
+
+TEST_F(SPGWAppInjectedStateProcedureTest, TestReleaseBearerWithInvalidImsi64) {
+  spgw_state_t* spgw_state        = get_spgw_state(false);
+  status_code_e return_code       = RETURNerror;
+  spgw_ue_context_t* ue_context_p = spgw_get_ue_context(test_imsi64);
+  teid_t ue_sgw_teid =
+      LIST_FIRST(&ue_context_p->sgw_s11_teid_list)->sgw_s11_teid;
+  s_plus_p_gw_eps_bearer_context_information_t* spgw_eps_bearer_ctxt_info_p =
```
Comment (ssanadhya) [inline lte/gateway/c/core/oai/test/spgw_task/test_spgw_procedures_with_injected_state.cpp]: Unused, please remove

**Review by ssanadhya — CHANGES_REQUESTED** [review]
Please cleanup unused variables.

**Review by ssanadhya — APPROVED** [review]

**Comment by ssanadhya** [comment]
@thhung , can you check why MME unit tests are failing?

**Comment by thhung** [comment]
> @thhung , can you check why MME unit tests are failing?

@ssanadhya : Unfortunately, even trying several times, I cannot reproduce these errors on my local testbed. 
It is related to the TestPagingMaxRetx of MME Unittest. There are 3 problems: 
1. after Modify Bearer Response, an attached UE is expected. However, there is no UE in the MME state.
2. sgw_handle_delete_session_request() is called twice when being called once is expected.
3. s1ap_handler->s1ap_handle_paging_request() is expected to be called twice but it was called only once.

**Comment by thhung** [comment]
> 



> @thhung , can you check why MME unit tests are failing?

Hi @ssanadhya , the test passed this time when I pushed a new commit. We might need to investigate more. It is hardly reproducible. I don't have this problem in my local test-bed.

**Comment by rsarwad** [comment]
Hi @ardzoht, may I know how is  file, IMSI001010000000001_S1AP_ATTACHED generated?
